### ISSUE TITLE
CockroachDB compatibility

### ISSUE THREAD

**Issue by zinid** [processone/ejabberd#3074]
ejabberd is mostly working with [CockroachDB](https://github.com/cockroachdb/cockroach), however several problems remain:
1. CockroachDB reserves `FAMILY`as a keyword. In the schema I fixed it by placing `family` in double quotes, however, the database still complains when `family` is put unquoted into SQL queries, but quoting it breaks MySQL compatibility (see #3072).
2. CockroachDB doesn't support `LIKE predicate ESCAPE clause` syntax (see E061-05 clause at https://www.cockroachlabs.com/docs/stable/detailed-sql-support.html). We use this syntax in some queries (in mod_pubsub for example). This can be fixed by removing [these two lines](https://github.com/processone/ejabberd/blob/19.09.1/src/ejabberd_sql.erl#L235-L236), however this will most likely break MSSQL compatibility (see https://docs.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?view=sql-server-ver15).
3. There is some problem with backslash escaping, e.g. see these two requests (generated by mod_mam_sql):
```
I191029 13:33:29.522174 64880 sql/exec_log.go:162  [n1,client=127.0.0.1:47197,user=root] 48183 exec "" {} "INSERT INTO archive(username, server_host, \"timestamp\", peer, bare_peer, xml, txt, kind, nick) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)" {$1:"e'test_slave!#$%^*()`~+-;_=[]{}|\\\\'", $2:"'pgsql.localhost'", $3:"1572356009500878", $4:"e'test_master!#$%^*()`~+-;_=[]{}|\\\\@pgsql.localhost/resource!@#$%^&*()\\'\"`~<>+-/;:_=[]{}|\\\\'", $5:"e'test_master!#$%^*()`~+-;_=[]{}|\\\\@pgsql.localhost'", $6:"e'<message xml:lang=\\'en\\' to=\\'test_slave!#$%^*()`~+-;_=[]{}|\\\\@pgsql.localhost/resource!@#$%^&amp;*()&apos;&quot;`~&lt;>+-/;:_=[]{}|\\\\\\' from=\\'test_master!#$%^*()`~+-;_=[]{}|\\\\@pgsql.localhost/resource!@#$%^&amp;*()&apos;&quot;`~&lt;>+-/;:_=[]{}|\\\\\\' id=\\'3330809881392138156\\' xmlns=\\'jabber:client\\'><body xml:lang=\\'en\\'>5</body></message>'", $7:"'5'", $8:"'chat'", $9:"''"} 9.296 1 ""
I191029 13:33:32.533616 64880 sql/exec_log.go:162  [n1,client=127.0.0.1:47197,user=root] 48184 exec "" {} "SELECT \"timestamp\", xml, peer, kind, nick FROM archive WHERE (username = e'test_slave!#$%^*()`~+-;_=[]{}|\\\\\\\\') AND (server_host = 'pgsql.localhost') ORDER BY \"timestamp\" ASC" {} 6.332 0 ""
```
The `INSERT` query works correctly, and I can see the row in the database, however, the second query (`SELECT`) returns nothing. This is most likely due to overquoting of backslashes.

**Comment by nordurljosahvida** [comment]
@zinid what's the current status of CDB compatbility? I don't see it in the docs, do i just do everything as if i were using postgres? Thanks

**Comment by licaon-kter** [comment]
@nordurljosahvida He doesn't work for P1 anymore.

**Comment by nordurljosahvida** [comment]
No way, sorry to hear that!

**Comment by nordurljosahvida** [comment]
@licaon-kter does anyone else know about cdb backend?

**Comment by prefiks** [comment]
I believe it mostly works with current version, at least it think it passes all test when using it as pg backend, but i am not 100% sure about it.

**Comment by mremond** [comment]
@nordurljosahvida Do not hesitate to give it a try and report your findings. Thanks !

**Comment by prefiks** [comment]
I fixed issue on cockroachdb in mod_vcard_sql module in commit: 1e456065f6895e83597f62ff0c9b3bd1797f7d08, there is still one test failing in mod_pubsub, i will try to diagnose it and fix it.

**Comment by prefiks** [comment]
I prepared couple more fixes for cockroachdb compatibility (f12ee28660b02bb72c36a416a8cca50c9fec6cf9, 0a88d03dc980cbf3a59fed437388d9b33567d30b, 2d707cc0d2bcf6da1b97cc2351ba5f9a661fba9c, 039d786e1f91a4c703edee7a532a77c0273b09a1, 92a09fdb713cb33f0409378161ed41dc85a11eb5), with those test suite passes without any errors reported (both with and without using prepared statements).

I think with this it is now ready for testing.

**Comment by nordurljosahvida** [comment]
great thank you so much!
